Update release process (#112)
XAITK/xaitk-saliency: docs/release_process.rst

Steps of the xaitk-saliency Release Process
===========================================
Three types of releases are expected to occur:
- major
- minor

> - patch

Reviewer: Is this supposed to render as an enumerated list? The different types of releases appear in-line in RTD.

Author: Oh darn, that was indeed supposed to be a bulleted list. Let me fix that.

Author: Done.

Create and Merge Version Update Branch
--------------------------------------

Major and Minor Releases
^^^^^^^^^^^^^^^^^^^^^^^^
Major and minor releases may add one or more trivial or non-trivial features
and functionalities.

1. Create a new branch off of the ``master`` named something like
   ``update-to-v{NEW_VERSION}``, where ``NEW_VERSION`` is the new ``X.Y``
   version.

   a. Use the ``scripts/update_release_notes.sh`` script to update the project
      version number, create ``docs/release_notes/v{NEW_VERSION}.rst``, and add a
      new pending release notes stub file.

      .. code-block:: bash

>          $ ./scripts/update_release_notes.sh minor

Reviewer: Is it clear that this script works also for major releases, by simply changing the argument passed from `minor` to `major`?

Reviewer: More so the latter, that it is clear here in the documentation that the script can be used for major, minor, and/or patch releases. It's not a major issue, so feel free to ignore (was mainly thinking about the failure mode where someone blindly copies the command).

Author: Valid, I can expand the example.

   b. Add a descriptive paragraph under the title section of
      ``docs/release_notes/v{NEW_VERSION}.rst`` summarizing this release.

2. Push the created branch to the upstream repository, not your fork (this is

>    an exception to the normal forky workflow).

Reviewer: ```suggestion
   an exception to the normal fork workflow).
```

Author: I thought I had an article in mind that called it "forky", but it turns out it called it "forking", so I'll go with that. A number of current sources use the "Forking Workflow" so I will stick with that (like [this](https://www.atlassian.com/git/tutorials/comparing-workflows/forking-workflow) and [this](https://docs.gitlab.com/ee/user/project/repository/forking_workflow.html)).

Reviewer: Sounds good, thanks!

3. Create a pull/merge request for this branch with ``release`` as the merge
   target. This is to ensure that everything passes CI testing before making
   the release. If there is an issue, then topic branches should be made and
   merged into this branch until the issue is resolved.

> 4. Get a positive review.

Reviewer: Should this be made more explicit?

```suggestion
4. Get an approving review.
```

Author: Sure, done.

> 5. Merge version bump branch into the ``release`` branch.

Reviewer: Is this the same as for patches? i.e.

```
5. Merge the pull/merge request into the ``release`` branch.
```

I find the patch instructions a bit simpler to parse (but could just be me).

Author: Oops, that item is also supposed to be symmetric between major/minor/patch. The line present in the patch version is what I meant to propagate up.

   .. code-block:: bash

      $ git fetch --all
      $ git checkout upstream/master
      $ git merge --log --no-ff upstream/release
      $ git push upstream master

7. Update version reference in the `XAITK/xaitk.github.io home page`_ to the
   new version.

Patch Release
^^^^^^^^^^^^^
A patch release should only contain fixes for bugs or issues with an existing
release.
No new features or functionality should be introduced in a patch release.
As such, patch releases should only ever be based on an existing release point
(git tag).

This list assumes we are creating a new patch release off of the *latest*
release version, i.e. off of the ``release`` branch.
If a patch release for an older release version is being created, see the
`Patching an Older Release`_ section.

1. Create a new branch off of the ``release`` branch named something like
   ``update-to-v{NEW_VERSION}``, where ``NEW_VERSION`` is the target ``X.Y.Z``,
   including the bump in the patch (``Z``) version component.

   a. Use the ``scripts/update_release_notes.sh`` script to update the project
      version number, create ``docs/release_notes/v{NEW_VERSION}.rst``, and add

>       a new pending release notes stub file. E.g.

Reviewer: E.g. may be extraneous:
```suggestion
      a new pending release notes stub file.
```

Author: Done.

      .. code-block:: bash

         $ ./scripts/update_release_notes.sh patch

   b. Add a descriptive paragraph under the title section of
      ``docs/release_notes/v{NEW_VERSION}.rst`` summarizing this release.

2. Push the created branch to the upstream repository, not your fork (this is

>    an exception to the normal forky workflow).

Reviewer: Pending change above, should also update here:
```suggestion
   an exception to the normal fork workflow).
```

Author: Done.

3. Create a pull/merge request for this branch with ``release`` as the merge
   target. This is to ensure that everything passes CI testing before making
   the release. If there is an issue, then topic branches should be made and
   merged into this branch until the issue is resolved.

> 4. Get a positive review.

Reviewer: Pending change above, should also update here:
```suggestion
4. Get an approving review.
```

Author: Done.

5. Merge the pull/merge request into the ``release`` branch.

6. Tag the resulting merge commit.
   See `Tag new version`_ below for how to do this.

7. As a repository administrator, merge the ``release`` branch into ``master``

>    locally and push the updated ``master`` to upstream. (Replace "upstream"
>    with your applicable remote name.)

Reviewer: Slightly different than the instructions in the major/minor release above, not sure it matters much:
```suggestion
   locally and push the updated ``master`` to upstream. (Replace "upstream"
   in the example below with your applicable remote name.)
```

Author: Good catch, I missed making that edit symmetric.

    Fixes
    -----

Patching an Older Release

> """""""""""""""""""""""""

Reviewer: Is this header meant to be slightly smaller in text size? It looks a bit different than the "Stub Pending Notes File" and "Tag new version" sections that it falls in-between (probably difference between using ^^^ vs. """).

Author: I actually meant to remove the "Stub Pending Notes File" section, but apparently missed doing that, so I will remove that now (that's now controlled by the helper script). The "Patching an Older Release" is intended to be a sub-section under "Patch Release".

Author: In following the sphinx sections guidance ([here](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections)), `"""` is listed as next in line after `^^^`.

When patching a major/minor release that is not the latest pair, a branch needs

> to be created based on release version being patched to integrate the specific

Reviewer: ```suggestion
to be created based on the release version being patched to integrate the specific
```

Author: Done.

patches into.
This branch should be prefixed with ``release-`` to denote that it is a release
integration branch.
Patch topic-branches should be based on this branch.
When all fix branches have been integrated, follow the `Patch Release`_ section
above, replacing ``release`` branch references (merge target) to be the
``release-...`` integration branch.
Step 6 should be to merge this release integration branch into ``release``
first, and *then* ``release`` into ``master``, if applicable (some patches may
only make sense for specific versions).

Tag new version
---------------
Release branches are tagged in order to record where in the git tree a
particular release refers to.
All release tags should be in the history of the ``release`` and ``master``
branches (barring exceptional circumstances).

> We will prefer to use local ``git tag`` commands to create the release version

Reviewer: ```suggestion
We prefer to use local ``git tag`` commands to create the release version
```

Author: Done

tag, pushing the tag to upstream.
The version tag should be applied to the merge commit resulting from the
above described ``update-to-v{NEW_VERSION}`` topic-branch ("the release").

See the example commands below, replacing ``HASH`` with the appropriate git
commit hash, and ``UPSTREAM`` with the appropriate remote name.

> We show using `Poetry's version command`_ to consistently access the current

Reviewer: ```suggestion
We show how to use `Poetry's version command`_ to consistently access the current
```

Author: Done
